python-backend/app/api/permits/permit/resources/permit.py

```diff
@@ -1,15 +1,102 @@
+
+    parser = reqparse.RequestParser()
+    parser.add_argument('permit_no', type=str, help='Number of the permit being added.')
+    parser.add_argument('mine_guid', type=str, help='guid of the mine.')
+    parser.add_argument(
+        'permit_status_code',
+        type=str,
+        help='Status of the permit being added.',
+        store_missing=False)
+    parser.add_argument(
+        'received_date', type=lambda x: datetime.strptime(x, '%Y-%m-%d') if x else None)
+    parser.add_argument(
+        'issue_date', type=lambda x: datetime.strptime(x, '%Y-%m-%d') if x else None)
+    parser.add_argument(
+        'authorization_end_date', type=lambda x: datetime.strptime(x, '%Y-%m-%d') if x else None)
+    parser.add_argument(
+        'permit_amendment_status_code', type=str, help='Status of the permit being added.')
+
     @api.doc(params={'permit_guid': 'Permit guid.'})
     @requires_role_mine_view
-    def get(self, permit_guid):
+    def get(self, permit_guid=None):
+        permit_no = request.args.get('permit-no', None, type=str)
+
+        if permit_no:
+            permit = Permit.find_by_permit_no(permit_no)
+            if permit:
+                return self.create_error_payload(
+                    400, 'There was a permit found with the provided permit number.'), 400
+
         permit = Permit.find_by_permit_guid(permit_guid)
+        if not permit:
+            return self.create_error_payload(404, 'Permit not found'), 404
+        return permit.json()
+
+    @api.doc(params={'permit_guid': 'Permit guid.'})
+    @requires_role_mine_create
+    def post(self, permit_guid=None):
+
+        if permit_guid:
+            return self.create_error_payload(400, 'unexpected permit_guid'), 400
+
+        data = self.parser.parse_args()
+
+        mine = Mine.find_by_mine_guid(data.get('mine_guid'))
+
+        if not mine:
+            return self.create_error_payload(404, 'There was no mine found with that guid.'), 404
+
+        permit = Permit.find_by_permit_no(data.get('permit_no'))
+
         if permit:
-            return permit.json()
-        return self.create_error_payload(404, 'Permit not found'), 404
+            return self.create_error_payload(
+                400, 'There was a permit found with the provided permit number.'), 400
+
+        permit = Permit.create(mine.mine_guid, data.get('permit_no'),
+                               data.get('permit_status_code'), self.get_create_update_dict())
+
+        amendment = PermitAmendment.create(permit,
+                                           data.get('received_date'), data.get('issue_date'),
+                                           data.get('authorization_end_date'),
+                                           self.get_create_update_dict(), 'OGP', 'ACT')
+
+        try:
+            amendment.save()
+            permit.save()
+        except AssertionError as e:
+            self.raise_error(500, 'Error: {}'.format(e))
```

Comment:
Nathan actually pointed this out, the assertion error could be thrown in the create helpers, it appears that once an object is attached to sql alchemy, the @validates are ran on assignment.

Reply:
K, I will change this to a basic exception.

Reply:
all the validate methods raise Assertion errors, but the try block needs to include the create methods.